vpr/src/place/initial_placement.cpp

```diff
@@ -549,47 +547,188 @@ static std::vector<ClusterBlockId> find_centroid_loc(const t_pl_macro& pl_macro,
+        loc_queue.pop();
+
+        // If this location has already been visited, skip it.
+        if (visited[loc.x][loc.y])
+            continue;
+        visited[loc.x][loc.y] = true;
+
+        // Get the distance from this loc to the src_loc in grid space
+        // Note: In compressed space, distances are not what they appear. We are
+        //       using the true grid positions to get the truly closest loc.
+        //       Here, we use the flat centroid of the macro and measure its
+        //       distance from the center of this grid loc (hence the 0.5 offset).
+        auto grid_loc = compressed_block_grid.compressed_loc_to_grid_loc(loc);
+        float grid_dx = std::abs((float)grid_loc.x + 0.5f - src_flat_loc.x);
+        float grid_dy = std::abs((float)grid_loc.y + 0.5f - src_flat_loc.y);
+        float grid_dist = grid_dx + grid_dy;
+        // If this distance is worst than the best we have seen or is outside
+        // of the search distance, do not explore it or its neighbors.
+        // NOTE: This prune is always safe (i.e. it will never remove a better
+        //       solution) since this is a spatial graph and our objective is
+        //       positional distance. The un-visitied neighbors of a node should
+        //       have a higher distance than the current node.
+        if (grid_dist >= best_dist)
+            continue;
+
+        // In order to ensure our BFS finds the closest compatible location, we
+        // traverse compressed grid locations which may not actually be valid
+        // (i.e. no tile exists there). This is fine, we just need to check for
+        // them to ensure we never try to put a cluster there.
+        bool is_valid_compressed_loc = false;
+        const auto& compressed_col_blk_map = compressed_block_grid.get_column_block_map(loc.x, 0);
+        if (compressed_col_blk_map.count(loc.y) != 0)
+            is_valid_compressed_loc = true;
+
+        // If this distance is better than the best we have seen so far, try
+        // to see if this is a better solution.
+        if (is_valid_compressed_loc) {
+            // Check if a sub-tile is available at this location.
+            const t_physical_tile_type_ptr phy_type = g_vpr_ctx.device().grid.get_physical_type(grid_loc);
+            const auto& compatible_sub_tiles = compressed_block_grid.compatible_sub_tile_num(phy_type->index);
+            int new_sub_tile = -1;
+            for (int sub_tile : compatible_sub_tiles) {
+                if (blk_loc_registry.grid_blocks().is_sub_tile_empty(grid_loc, sub_tile)) {
+                    new_sub_tile = sub_tile;
+                    break;
+                }
+            }
+            if (new_sub_tile != -1) {
+                // If a sub-tile is available, set this to be the first sub-tile
+                // available and check if this site is legal for this macro.
+                // TODO: Should we pick a random sub-tile instead?
```

Comment:
First should be fine ... there's no spatial meaning of difference between the subtiles at a certain (x,y,layer).